README.md
## Monty Host Inspector (MHI) (v1.0.3)
Version: 1.0.3

Information:
- Montybot is a work in progress ChatOps tool to help enumerate subdomains and/or to fuzz directories of given websites or web applications. 

Dependencies:
- ghostbin.py (simple ghostbin request post module)
- [matrix_bot_api](https://github.com/shawnanastasio/python-matrix-bot-api) (modified to include whitelisting)
- [sublist3r](https://github.com/aboul3la/Sublist3r)
- [wfuzz](https://github.com/xmendez/wfuzz)
- wordlists from: 
  - https://github.com/henshin/filebuster/tree/master/wordlists
  - https://github.com/danielmiessler/SecLists/blob/master/Miscellaneous/wordlist-skipfish.fuzz.txt

Requirements:
- config.py:
```python
USERNAME = ""  # Bot's username, in the format of `@user:server.com`
PASSWORD = ""  # Bot's password
SERVER = ""  # Matrix server URL, in the format of `https://server.com`
GITHUB_TOKEN = "" # Github API token for posting Gists
```
- user_whitelist.txt:
```text
@username:matrix.org
@username:custom.server.org
```

Usage instructions:
- **!helpmonty**: display a help message
- **!hello**: sends a greeting to the user
- **!hi https://example.com**: inspects a host with wfuzz. command incorporates *power* settings (1-4), which increase the size/strength of the wordlists. increase with **!hi https://example.com override (1-4)**
- **!si example.com**: enumerates subdomains with sublist3r (bruteforce capability and engines set to off/fast)


CHANGELOG.md
# Changelog: 

### 1.0.3
**Feature enhancements:**

- Added support for GitHub Gists. Previously used data uploading sites (pastebin, hastebin, ghostbin,etc) were insufficient for the amount of data that could be accumulated via host or subdomain inspections. Create an API with gist permission, and add your API token in `config.py`. Due to some character limitations in Gist creation, URLs are SHA1 hashed, and the resultant hash is used as the Gist title. 
- Added a percentage based status indicator for subdomain inspection. Monty will send an update at approximately 25, 50 and 75 percent complete (not rounded because it looks cool)

**Bugfixes:**

- Added logic to fix an infinite redirect for sites which redirect invalid paths to a homepage or other landing page. 
- Added and removed support for Sprunge.us as a data uploader. 
- Added user whitelisting for the basic POST callback, `!post`.
- Added basic URL input sanitization for subdomain inspection. Subdomain inspection now supports `https://google.com/` as input, or `google.com` (`https://google.com///////////' turns into `https://google.com/`)
- General text cleanup/formatting.

### 1.0.2
**Feature enhancements:**

- Added basic web port (80/443) inspection for enumerated hosts with a chat status update.
- Added basic POST callback `!post` to run through the startup commands for debugging purposes. 

**Bugfixes:**

- Updated help menu.
- Added error handling for hastebin uploads. 
- Added checks to validate override parameters for host inspections.
- Fixed error with host inspection not properly handling integer return codes.
- Added more verbosity to errors provided by host inspection. 
- Migrated hastebin.com data upload to ghostbin.co due to 403 errors returned.

### 1.0.1
- Modified wfuzz calls to follow redirects to end, which removes the necessity for a percentage based result filter. (initial commit)
- Hardcoded `!hi` to have a strength of 1 by default, for easier usage. Added override to allow more detailed wordfilters, updated console debugging. (initial commit)

### 1.0.0 
Initial release. (initial commit)


matrix_bot_api/matrix_bot_api.py
import traceback
import re
from matrix_client.client import MatrixClient
from matrix_client.api import MatrixRequestError

class MatrixBotAPI:
    # username - Matrix username
    # password - Matrix password
    # server   - Matrix server url : port
    # rooms    - List of rooms ids to operate in, or None to accept all rooms
    def __init__(self, username, password, server, rooms=None):
        self.username = username

        # Authenticate with given credentials
        self.client = MatrixClient(server)
        try:
            self.client.login_with_password(username, password)
        except MatrixRequestError as e:
            print(e)
            if e.code == 403:
                print("Bad username/password")
        except Exception as e:
            print("Invalid server URL")
            traceback.print_exc()

        # Store allowed rooms
        self.rooms = rooms

        # Store empty list of handlers
        self.handlers = []

        # If rooms is None, we should listen for invites and automatically accept them
        if rooms is None:
            self.client.add_invite_listener(self.handle_invite)
            self.rooms = []

            # Add all rooms we're currently in to self.rooms and add their callbacks
            for room_id, room in self.client.get_rooms().items():
                room.add_listener(self.handle_message)
                self.rooms.append(room_id)
        else:
            # Add the message callback for all specified rooms
            for room in self.rooms:
                room.add_listener(self.handle_message)

    def add_handler(self, handler):
        self.handlers.append(handler)

    def handle_message(self, room, event):
        # Make sure we didn't send this message
        if re.match("@" + self.username, event['sender']):
            return

        # Loop through all installed handlers and see if they need to be called
        for handler in self.handlers:
            if handler.test_callback(room, event):
                # This handler needs to be called
                try:
                    handler.handle_callback(room, event)
                except:
                    traceback.print_exc()

    def handle_invite(self, room_id, state):
        inviting_user = state['events'][1]['sender']
        print("Got invite to room: {} by {}.".format(str(room_id), inviting_user))
        
        # Check if we're allowed to be invited by the whitelist
        with open('user_whitelist.txt') as file:
            users = file.readlines()
        
        if [user for user in users if re.match(inviting_user, user)]:
            print("Accepting invite from {}. Joining room.".format(inviting_user))
            room = self.client.join_room(room_id)
            
            # Add message callback for this room
            room.add_listener(self.handle_message)
            
            # Add room to list
            self.rooms.append(room)
        else: 
            print("Not accepting invite from: {}. Add to user_whitelist.txt to enable access.".format(inviting_user))
    
    def set_matrix_avatar(self, url):
        avatar = MatrixClientUser.set_avatar_url(self, url)
        return avatar

    def start_polling(self):
        # Starts polling for messages
        self.client.start_listener_thread()
        return self.client.sync_thread


modified_dependencies/github_gist.py
import requests
import json

def post(api_token, description, visibility, content):
    api_url = 'https://api.github.com/gists'
    token = 'token {}'.format(api_token)
    
    # Set the visibility to public if preferred, otherwise set to private
    if visibility.lower() == 'public':
        visibility = True
    elif visibility.lower() == 'private':
        visibility = False
    else:
        visibility = False
    
    headers = {'Authorization': token}
    params = {'scope': 'gist'}
    payload = {'description': description, 'public': visibility, 'files': {description: {'content': content}}}
    
    try:
        post = requests.post(api_url, headers=headers, data=json.dumps(payload))
        response = json.loads(post.text)
        return response['html_url']
    except Exception as e:
        return str(e)

bot.py
import sys, os, re, random, hashlib
import wfuzz
import config
import modified_dependencies.github_gist as github_gist

from scapy.all import * 
from random import randint
from Sublist3r.subbrute import subbrute
from Sublist3r import sublist3r
from matrix_bot_api.matrix_bot_api import MatrixBotAPI
from matrix_bot_api.mregex_handler import MRegexHandler
from matrix_bot_api.mcommand_handler import MCommandHandler

# Global variables
USERNAME = config.USERNAME
PASSWORD = config.PASSWORD
SERVER = config.SERVER
GITHUB_TOKEN = config.GITHUB_TOKEN

# inspection functions 
def host_inspection(host, power):
    # light, fast, normal, thorough are power choices
    fuzzable_url = str(os.path.join(str(host), 'FUZZ'))
    power = int(power)
    print('trying to fuzz: {}'.format(fuzzable_url))
    if power == 1:
        filename = 'wl/light.txt'
    elif power == 2:
        filename = 'wl/fast.txt'
    elif power == 3:
        filename = 'wl/normal.txt'
    elif power == 4:
        filename = 'wl/thorough.txt'
    else: 
        error = 'Incorrect power provided.'
        return error
    
    host_data = []
    try:
        url_placeholder = []
        for r in wfuzz.fuzz(url=fuzzable_url, follow=True, hc=[404], payloads=[('file', dict(fn=filename))]):
            # Basic checking for redirects such as ones on login forms
            if r.url in url_placeholder:
                continue
            url_placeholder.append(r.url)
            a = [r.code, r.url]
            host_data.append(a)
    except Exception as e:
        error = str(e)
        return error
    
    # Check to see if the results are somewhat valid by seeing if the first n results match the first n from the wordlist, otherwise return something else
    n = 0 
    host_count = len(host_data)
    sample_count = round(host_count/5)
    host_samples = host_data[:sample_count]
    with open(filename) as file:
        challenge_lines = file.readlines()[0:sample_count]
    
    for i in range(0,sample_count-1):
        challenge = re.compile('/' + str(challenge_lines[i].strip('\n')) + '$')        
        
        for host in host_samples:
            if challenge.findall(host[1]):
                n += 1
    
    print('total matches: {}'.format(str(n)))
    print('total match percent: {}/{}'.format(n, sample_count))
    
    if n > sample_count/2: 
        # if more than half the results are trash we are assuming that it has a 404 redirect
        error = 'Match sample count failed check.'
        return error
    
    return host_data

def subdomain_inspection(host, room):
    # hardcoding engine power for now
    engine_power = 'light'
    
    # engine power: light, medium and very slow 
    if engine_power == str('light'):
        engine_list = 'google,yahoo,bing,ask,netcraft,virustotal,threatcrowd'
        bf = False
    elif engine_power == str('medium'):
        engine_list = 'google,yahoo,bing,ask,netcraft,dnsdumpster,virustotal,threatcrowd'
        bf = False
    elif engine_power == str('very_slow'):
        engine_list = 'google,yahoo,bing,ask,netcraft,dnsdumpster,virustotal,threatcrowd,baidu'
        bf = True
        # print that this is going to take forever
    else:  
        return False
    
    # enumerate subdomain information 
    try:
        domain_results = sublist3r.main(host, 200, None, ports=None, silent=True, verbose=False, enable_bruteforce=bf, engines=engine_list)
        print(domain_results)
        # return results
    except Exception as e:
        return str(e)
    
    port_results = []
    ports_to_test = [80,443]
    # try to do some basic port checking 
    room.send_html('<blockquote>Subdomain enumeration complete for {}. Port inspection initiated.</blockquote>'.format(host))
    print('- performing subdomain port inspection for host {}'.format(host))
    total_results = len(domain_results)
    current_iteration = 0
    percent_flag = 0
    
    for result in domain_results: 
        port_result_holder = []
        
        port = randint(27000,28000)
        for port in ports_to_test:
            try:
                ps = IP(dst=result)/TCP(sport=port,dport=port)
                sr = sr1(ps,timeout=1,verbose=0)
                if sr:
                    port_result_holder.append(str(port))
                else:
                    port_result_holder.append('0')
            except Exception as e:
                print('port check error: {}'.format(str(e)))
                port_result_holder.append('0')
        
        port_results.append(port_result_holder)
        
        current_percent = (current_iteration/total_results)*100
        rounded_percent = 5 * round(current_percent/5)
        if (int(rounded_percent) == 25) and (percent_flag == 0):
            room.send_html('<pre><code>{} percent complete: {}%.</pre></code>'.format(str(host), str(current_percent)))
            percent_flag += 1
        elif (int(rounded_percent) == 50) and (percent_flag == 1):
            room.send_html('<pre><code>{} percent complete {}%.</pre></code>'.format(str(host), str(current_percent)))
            percent_flag += 1
        elif (int(rounded_percent) == 75) and (percent_flag == 2):
            room.send_html('<pre><code>{} percent complete {}%.</pre></code>'.format(str(host), str(current_percent)))
            percent_flag += 1 
        
        print('Rounded {}, Current {}, % flag {}'.format(rounded_percent, current_percent, percent_flag))
        current_iteration += 1
    
    print('- subdomain port inspection completed for host {}'.format(host))
    # combine the lists
    results = {}
    results['domains'] = domain_results
    results['port_results'] = port_results
    
    return results

def hostname_cleanup(hostname, callback):
    if callback == 'si':
        pattern = re.compile(r'(http|https)://')
        hostname = pattern.sub('', hostname).rstrip('/')
        return hostname
    elif callback == 'hi':
        # nothing yet
        return hostname 

# callbacks
def help_callback(room, event):
    room.send_html("""
                    <h4 style="margin-bottom:0px">Monty Host Inspector (MHI)</h4>
                    <b>Version: 1.0.3</b><br />
                    <b>Information: </b><br />
                    <ul style="list-style=none">
                        <li>I can help you to enumerate subdomains and/or to fuzz directories of given websites or web applications.</li>
                        <li>It takes me time to return information! Please be patient.</li>
                    </ul>    
                    <b>Dependencies: </b><br />
                    <ul>
                        <li>github_gist.py (requests wrapper for posting to your github)</li>
                        <li>matrix_bot_api (modified to include whitelisting)</li>
                        <li>sublist3r</li>
                        <li>scapy</li>
                        <li>wfuzz</li>
                        <li>wordlists from: 
                            <ul>
                                <li>https://github.com/henshin/filebuster/tree/master/wordlists</li>
                                <li>https://github.com/danielmiessler/SecLists/blob/master/Miscellaneous/wordlist-skipfish.fuzz.txt</li>
                            </ul>
                        </li>
                    </ul>
                    <b>Usage instructions: </b><br />
                    <ul>
                        <li><b>!helpmonty</b>: this message</li>
                        <li><b>!hello</b>: sends a greeting to the user</li>
                        <li><b>!hi https://example.com</b> (defaults power to 1, increase with <b>!hi https://example.com override (1-4)</b></li>
                        <li><b>!si example.com</b>: enumerates subdomains with sublist3r (bf and engine power set to off/1)</li>
                    </ul>
                    """)
           
def hello_callback(room, event):
    # Somebody said hi, let's say Hi back
    room.send_html("<pre>Hello, " + event['sender'] + '</pre>')

def hi_callback(room, event):
    pattern = re.compile('^(https|http)\:\/\/(.*)')
    args = event['content']['body'].split()
    args.pop(0)

    if not (len(args) != 1 or len(args) != 3):
        # tell them some info
        room.send_html('too many or too few arguments for mon_ty to process. see <b>!helpmonty</b> for usage')
        return False
    else:
        if len(args) == 3:
            # make arguments readable
            hostname = args[0]
            override = args[1]
            power = args[2] 
            
            if override != 'override': 
                room.send_html('second argument is an override parameter. see <b>!helpmonty</b> for more info')
                return False
            
            if not isinstance(int(power), int):
                room.send_html('third argument must be an int ranging from 1-4. see <b>!helpmonty</b> for more info')
                return False
            
            if 1 < int(power) > 4:
                room.send_html('third argument must be between 1-4. see <b>!helpmonty</b> for more info')
                return False                
            
        elif len(args) == 1:
            hostname = args[0]
            power = 1
        
        if not pattern.findall(hostname): 
            room.send_text('improperly formatted url. please provide a url like http://google.com or https://google.com. Thank you.')
            return False
        
        else: 
            print('Inspecting {}.'.format(hostname))
            data = host_inspection(hostname, power)
            if isinstance(data, str):
                room.send_html('Error occurred during fuzzing. <pre><code>{}</code></pre>'.format(data))
            else: 
                if len(data) > 1:
                    gist_results = 'Code: Returned record'.format(hostname)
                    formatted_results = '<pre><code><b>Code: Returned record</b>'
                    for item in data:
                        if item[0] == 200:
                            port_data = '\n<b>{}</b> : {}'.format(item[0], item[1])
                        else:
                            port_data = '\n{} : {}'.format(item[0], item[1])
                        gist_results += '\n{} : {}'.format(item[0], item[1])
                        formatted_results += port_data
                        
                    # Gist an unformatted version, sha1 hash the hostname because you can't use forward slashes in github gist titles
                    try:
                        formatted_hostname = hashlib.sha1(hostname.encode('utf-8')).hexdigest()
                        data_post_url = github_gist.post(GITHUB_TOKEN, '{} hi data'.format(formatted_hostname), 'Private', gist_results)
                        room.send_html('Results for <b>{} (id:{})</b> (view @ {}):'.format(hostname, formatted_hostname, data_post_url))
                    except:
                        room.send_html('Results for <b>{}</b> (data upload error):'.format(hostname))
                        print(str(e))
                    
                    try:
                        print(formatted_results)
                        room.send_html('{}\n</code></pre>'.format(formatted_results))
                    except:
                        data_post_url = github_gist.post(GITHUB_TOKEN, '{} hi data'.format(hostname), 'Private', formatted_results)
                        room.send_text('Request too large for Matrix to handle. View results at {}'.format(data_post_url))
                else: 
                    room.send_html('Unable to find any results for: {}'.format(hostname))

def si_callback(room, event):
    print(event)
    args = event['content']['body'].split()
    args.pop(0)
    
    if len(args) != 1:
        # tell them some info
        room.send_html('si command takes ONE, count it, one, argument. see <b>!helpmonty</b> for more info')
        return
    
    hostname = args[0]
    print('subdomain inspection callback initiated for {}'.format(hostname))
    # do error handling on this later 
    hostname = hostname_cleanup(hostname, 'si')
    results = subdomain_inspection(hostname, room)
    
    if len(results['domains']) == 0:
        room.send_html('<pre>Domain formatted incorrectly</pre>')
        return
    if len(results) > 0: 
        # parse the results 
        i = 0
        formatted_results = []
        for domain in results['domains']:
            # join the domains with the ports
            webports = ''
            ports = results['port_results'][i]
            for port in ports:
                if port != '0':
                    # append the port
                    if len(webports) == 0:
                        webports = port
                    else: 
                        webports = webports + ' ' + port
            
            domain_result = str('Domain: {} || Webports: {}'.format(domain, webports))
            i += 1
            
            formatted_results.append(domain_result)
        
        # upload formatted_results to paste host and dump an error if fails
        try:
            print(formatted_results)
            data_post_url = github_gist.post(GITHUB_TOKEN, '{} si data'.format(hostname), 'Private', '\n'.join(formatted_results))
            room.send_html('Results for <b>{}</b> (view @ {}):'.format(hostname, data_post_url))
        except Exception as e:
            room.send_html('Results for <b>{}</b> (data upload error):'.format(hostname))
            print('data upload error for domain {}'.format(hostname))
            print(str(e))
        
        try: 
            room.send_html('<pre><code>{}</code></pre>'.format('\n'.join(formatted_results)))
            print('subdomain inspection callback completed for {}'.format(hostname))
        except Exception as e:
            truncated_results = formatted_results[:150]
            room.send_text('Truncating to 150 results.')
            room.send_html('<pre><code>{}</code></pre>'.format('\n'.join(truncated_results)))
            print('subdomain inspection callback failed for {}'.format(hostname))
            print('reason: {}'.format(str(e)))
    
    return

def post_callback(room, event):
    requesting_user = event['sender']
    with open('user_whitelist.txt') as file:
            users = file.readlines()
    
    if [user for user in users if re.match(requesting_user, user)]:
        fake_subdomain_event = {'event_id': 'abcdefg', 'sender': 'sender', 'origin_server_ts': 1234567890, 'content': {'msgtype': 'm.text', 'body': '!si matrix.org'}, 'room_id': '!abcdefg', 'unsigned': {'age': 99}, 'type': 'm.room.message'}
        fake_host_event = {'event_id': 'abcdefg', 'sender': 'sender', 'origin_server_ts': 1234567890, 'content': {'msgtype': 'm.text', 'body': '!hi https://matrix.org'}, 'room_id': '!abcdefg', 'unsigned': {'age': 99}, 'type': 'm.room.message'}
        room.send_html('<pre>Initiated callback POST with limited reporting.</pre>')
        
        room.send_html('<pre><code>Performing POST subdomain callback.</pre></code>')
        try: 
            si_callback(room, fake_subdomain_event)
        except Exception as e: 
            print('<pre><code>SI callback error: {}</code></pre>'.format(str(e)))
        
        room.send_html('<pre><code>Performing POST host inspection callback.</pre></code>')
        try: 
            hi_callback(room, fake_host_event)
        except Exception as e: 
            print('<pre><code>HI callback error: {}</code></pre>'.format(str(e)))
        
        room.send_html('<pre><code>Callback POST complete.</code></pre>') 
        return True
    else: 
        room.send_html('Sorry {}, you do not have the authorization to run this command!'.format(requesting_user))
        return True

def main():
    # Create an instance of the MatrixBotAPI
    bot = MatrixBotAPI(USERNAME, PASSWORD, SERVER)

    # Set avatar / fix this later
    # bot.set_matrix_avatar(AVATAR_URL)
    
    # Hello testing handler 
    hello_handler = MCommandHandler("hello", hello_callback)
    bot.add_handler(hello_handler)
    
    # Host inspection handler
    hi_handler = MCommandHandler("hi", hi_callback)
    bot.add_handler(hi_handler)
    
    # Subdomain inspection handler 
    si_handler = MCommandHandler("si", si_callback)
    bot.add_handler(si_handler)
    
    # Help handler
    help_handler = MCommandHandler("helpmonty", help_callback)
    bot.add_handler(help_handler)
    
    # POST handler
    post_handler = MCommandHandler("post", post_callback)
    bot.add_handler(post_handler)
    
    # Start polling
    bot.start_polling()
    
    # Infinitely read stdin to stall main thread while the bot runs in other threads
    while True:
        input()

if __name__ == "__main__":
    main()

